src/micro_endpoint.c

```diff
@@ -0,0 +1,471 @@
+// Copyright 2021-2023 The NATS Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <ctype.h>
+
+#include "microp.h"
+#include "mem.h"
+
+static microError *dup_with_prefix(char **dst, const char *prefix, const char *src);
+static void free_endpoint(microEndpoint *ep);
+static void handle_request(natsConnection *nc, natsSubscription *sub, natsMsg *msg, void *closure);
+static microError *new_endpoint(microEndpoint **new_ep, microService *m, const char *prefix, microEndpointConfig *cfg, bool is_internal);
+static void on_drain_complete(void *closure);
+static void release_endpoint(microEndpoint *ep);
+static microEndpoint *retain_endpoint(microEndpoint *ep);
+
+microError *
+micro_new_endpoint(microEndpoint **new_ep, microService *m, const char *prefix, microEndpointConfig *cfg, bool is_internal)
+{
+    microError *err = NULL;
+    microEndpoint *ep = NULL;
+    const char *subj;
+
+    if (!micro_is_valid_name(cfg->Name) || (cfg->Handler == NULL))
```

Comment:
Should we check that `cfg` is not NULL here too?

Reply:
done